packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -236,52 +237,81 @@ async function resolveStaticMetadata(components: ComponentsType) {
+  const resolvers: ((value: ResolvedMetadata) => void)[] = []
+  const generateMetadataResults: (Metadata | Promise<Metadata>)[] = []
 
-  for (const item of metadataItems) {
-    const [metadataExport, staticFilesMetadata] = item
-    const currentMetadata =
-      typeof metadataExport === 'function'
-        ? metadataExport(parentPromise)
-        : metadataExport
-    const layerMetadataPromise =
-      currentMetadata instanceof Promise
-        ? currentMetadata
-        : Promise.resolve(currentMetadata)
+  // call each `generateMetadata function concurrently and stash their resolver
+  for (let i = 0; i < metadataItems.length; i++) {
+    const { metadataExport } = metadataItems[i]
+    if (typeof metadataExport === 'function') {
+      generateMetadataResults.push(
+        metadataExport(
+          new Promise((resolve) => {
+            resolvers.push(resolve)
+          })
+        )
+      )
+    }
+  }
 
-    parentPromise = parentPromise.then((resolved) => {
-      return layerMetadataPromise.then((metadata) => {
-        resolved = deepClone(resolved)
-        merge(resolved, metadata, staticFilesMetadata, {
-          title: resolved.title?.template || null,
-          openGraph: resolved.openGraph?.title?.template || null,
-          twitter: resolved.twitter?.title?.template || null,
-        })
-        return process.env.NODE_ENV === 'development'
-          ? Object.freeze(resolved)
-          : resolved
-      })
+  // Loop over all metadata items again, merging synchronously any static object exports,
+  // awaiting any static promise exports, and resolving parent metadata and awaiting any generated
+  // metadata
+  let resolvingIndex = 0
+  for (let i = 0; i < metadataItems.length; i++) {
+    const { metadataExport, staticFilesMetadata, isLeafLayout } =
+      metadataItems[i]
+    let metadata: Metadata | null = null
+    if (typeof metadataExport === 'function') {
+      const resolveParent = resolvers[resolvingIndex]
+      const generatedMetadata = generateMetadataResults[resolvingIndex]
+      resolvingIndex++
+      // In dev we clone and freeze to prevent relying on mutating resolvedMetadata directly.
+      // In prod we just pass resolvedMetadata through without any copying.
+      const currentResolvedMetadata: ResolvedMetadata =
+        process.env.NODE_ENV === 'development'
+          ? Object.freeze(
+              require('next/dist/compiled/@edge-runtime/primitives/structured-clone').structuredClone(
+                resolvedMetadata
+              )
+            )
+          : resolvedMetadata
+
+      // This resolve should unblock the generateMetadata function if it awaited the parent
+      // argument. If it didn't await the parent argument it might already have a value since it was
+      // called concurrently. Regardless we await the return value before continuing on to the next
+      // layer
+      resolveParent(currentResolvedMetadata)
+      metadata = await generatedMetadata
```

Comment:
We should probably check whether generateMetadata is a Promise and only resolve and await if it is. If it is a regular object we can just treat it like a sync metadata and skip the extra microtask